src/webgpu/api/validation/copyTextureToTexture.spec.ts

```diff
@@ -210,3 +236,222 @@ g.test('multisampled_copy_restrictions')
+      usage: GPUTextureUsage.COPY_SRC,
+    });
+
+    const dstTexture = t.device.createTexture({
+      size: kTextureSize,
+      format: dstFormat,
+      usage: GPUTextureUsage.COPY_DST,
+    });
+
+    const isSuccess = srcFormat === dstFormat;
+    t.TestCopyTextureToTexture(
+      { texture: srcTexture },
+      { texture: dstTexture },
+      kTextureSize,
+      isSuccess
+    );
+  });
+
+g.test('depth_stencil_copy_restrictions')
+  .params(
+    params()
+      .combine(poptions('format', kDepthStencilFormats))
+      .combine(
+        poptions('srcCopyOrigin', [
+          { x: 0, y: 0, z: 0 },
+          { x: 1, y: 0, z: 0 },
+          { x: 0, y: 1, z: 0 },
+          { x: 1, y: 1, z: 0 },
+        ])
+      )
+      .combine(
+        poptions('dstCopyOrigin', [
+          { x: 0, y: 0, z: 0 },
+          { x: 1, y: 0, z: 0 },
+          { x: 0, y: 1, z: 0 },
+          { x: 1, y: 1, z: 0 },
+        ])
+      )
+      .combine(poptions('srcTextureSize', [64, 32]))
+      .combine(poptions('dstTextureSize', [64, 32]))
+      .combine(poptions('srcCopyLevel', [1, 2]))
+      .combine(poptions('dstCopyLevel', [0, 1]))
+      .expand(p =>
+        poptions('copyWidth', [
+          Math.max(p.srcTextureSize >> p.srcCopyLevel, p.dstTextureSize >> p.dstCopyLevel) -
+            Math.max(p.srcCopyOrigin.x, p.dstCopyOrigin.x),
+          8,
+        ])
+      )
+      .expand(p =>
+        poptions('copyHeight', [
+          Math.max(p.srcTextureSize >> p.srcCopyLevel, p.dstTextureSize >> p.dstCopyLevel) -
+            Math.max(p.srcCopyOrigin.y, p.dstCopyOrigin.y),
+          8,
+        ])
+      )
+  )
+  .fn(async t => {
+    const {
+      format,
+      srcCopyOrigin,
+      dstCopyOrigin,
+      srcTextureSize,
+      dstTextureSize,
+      srcCopyLevel,
+      dstCopyLevel,
+      copyWidth,
+      copyHeight,
+    } = t.params;
+
+    const kMipLevelCount = 3;
+    const srcTexture = t.device.createTexture({
+      size: { width: srcTextureSize, height: srcTextureSize, depth: 1 },
+      format,
+      mipLevelCount: kMipLevelCount,
+      usage: GPUTextureUsage.COPY_SRC,
+    });
+    const dstTexture = t.device.createTexture({
+      size: { width: dstTextureSize, height: dstTextureSize, depth: 1 },
+      format,
+      mipLevelCount: kMipLevelCount,
+      usage: GPUTextureUsage.COPY_DST,
+    });
+
+    const isSuccess =
+      copyWidth === srcTextureSize >> srcCopyLevel &&
+      copyHeight === srcTextureSize >> srcCopyLevel &&
+      copyWidth === dstTextureSize >> dstCopyLevel &&
+      copyHeight === dstTextureSize >> dstCopyLevel;
+    t.TestCopyTextureToTexture(
+      { texture: srcTexture, origin: srcCopyOrigin, mipLevel: srcCopyLevel },
+      { texture: dstTexture, origin: dstCopyOrigin, mipLevel: dstCopyLevel },
+      { width: copyWidth, height: copyHeight, depth: 1 },
+      isSuccess
+    );
+  });
+
+g.test('copy_ranges')
+  .params(
+    params()
+      .combine(
+        poptions('srcCopyOrigin', [
```

Comment:
Maybe the same trick with offset could be used so that we test exactly at the correct boundary.

Reply:
I'm not sure I have fully understood your comments here. PTAL, thanks!